### ISSUE TITLE
swarm: Nodes count is off in `docker info` when worker leaves and rejoins

### ISSUE THREAD

**Issue by tiborvass** [docker/docker#24088]
```
manager$ docker swarm init
worker$ docker swarm join manager:2377
manager$ docker info | grep Nodes
 Nodes: 2
worker$ docker swarm leave
manager$ docker info | grep Nodes
 Nodes: 2
worker$ docker join manager:2377
manager$ docker info | grep Nodes
 Nodes: 3 
```

@tonistiigi @diogomonica is the issue that the new certs are being generated? Is it safe to reuse the certs that were there when the worker rejoins? Or should the worker tell the manager that it's leaving?

**Comment by tonistiigi** [comment]
`docker swarm leave` removes all current swarm state and gets you into same state that you were before `init/join`. If you do `node ls` then all these nodes are correctly marked as down and can be removed. Maybe we should not consider these in node count in `docker info`.

Reusing a certificate is probably possible but I don't see it being worth it. We have to check that it matches CA and isn't expired, plus maybe you don't want certificates/keys that can be used to connect to your cluster in the machine that doesn't use swarm at all.

**Comment by diogomonica** [comment]
In the future we could allow a node to delete itself once it gets a leave command, but this is not a bug and like @tonistiigi said, is side-effect of new certificates being issued for the new node.

**Comment by thaJeztah** [comment]
Do nodes have an identifier that doesn't change upon leave/join that we could do to remove the duplicates in the output?

**Comment by diogomonica** [comment]
@thaJeztah no, that is the wrong way of solving this issue. No other ID can be trusted.

**Comment by tiborvass** [comment]
@diogomonica how is this not a bug? There are 2 nodes in my swarm, and it says there are 3.

**Comment by diogomonica** [comment]
/cc @aaronlehmann 

@tiborvass it's not a bug because it was intentional. The only way to remove a node from the cluster after you doing is by doing `node rm` on a manager. I think we should do graceful removal on `docker swarm leave` though. That would be more user-friendly.

**Comment by tiborvass** [comment]
@diogomonica yes I guess that's what I'd like to see, so that the example I shown in the description shows `Nodes: 2` then `Nodes: 1`, then `Nodes: 2`.

**Comment by diogomonica** [comment]
Agreed.

**Comment by allencloud** [comment]
Any update here?

>  I think we should do graceful removal on docker swarm leave though

👍

**Comment by JokerQyou** [comment]
It's been 14 months since this issue was created, any progress?

**Comment by thaJeztah** [comment]
I don't think someone is working on this, and this probably needs changes in SwarmKit; contributions are welcome though 👍

**Comment by bachner** [comment]
Do you have any workaround for this issue?

**Comment by MarianaNRC** [comment]
I had a similar problem with a swarm consisting of one manager and a worker. Then the STATUS of the worker was down, so i just removed the worker node. Then created another one. The problem that 'docker node list' showed the old worker too. When i tried to remove it from the manager node , it gave me this error ' node OS-worker  is ambiguous (2 matches found)' 
The only solution i found was to remove the nodes from the manager node by ID instead by name. I used 'docker-machine ssh OS-manager "docker node remove -f beqndspacwoubee9g50s8fomm ". Where 'beqndspacwoubee9g50s8fomm' is the node ID. Thank God it worked finally....

**Comment by jrab66** [comment]
i would add if you try to use terraform to automate the swarm, removing or scaling workers leave the worker name with ambiguity and you have to remove it via id or some yanky script!
